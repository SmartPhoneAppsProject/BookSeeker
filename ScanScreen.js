import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  Image,
  Dimensions,
} from 'react-native';
import {
  BarCodeScanner,
  Permissions
} from 'expo';
import { Button } from 'react-native-elements';

import { postBook } from './Network';

export default class ScanScreen extends Component {
  static navigationOptions = {
    title: 'バーコードリーダー',
  };

  constructor(props) {
    super(props);
    this.state = {
      janCode: null,
      permissionsGranted: false,
      status: 'reading',
    };
  }

  async componentWillMount() {
    const { status } = await Permissions.askAsync(Permissions.CAMERA);
    this.setState({ permissionsGranted: status === 'granted' });
  }

  _handleBarCodeRead = ({ type, data }) => {
    console.log(data);
    if (`${BarCodeScanner.Constants.BarCodeType.ean13}` == type) {
      if (978 == data.slice(0, 3)) { //ISBNを読み取ったとき
        if (this.state.janCode != data) {
          this.setState({
            janCode: data,
            status: 'ok'
          });
          setTimeout(() => {
            const janCode = parseInt(data);
            this.registerBook(janCode);
          }, 1000);
        }
      } else { //バーコードであるがISBNでないとき
        this.setState({ status: 'invalid' });
      }
      setTimeout(() => {
        this.setState({ status: 'reading' });
      }, 1000);
    } else { //バーコードでないとき
      this.setState({ status: 'reading' });
    }
  };

  registerBook = (janCode) => {
    const { navigate } = this.props.navigation;
    const { params } = this.props.navigation.state;

    const json = JSON.stringify({
      title: params.title,
      // image: params.photo.base64,
      image: '',
      published_at: params.publishedAt,
      jan_code: janCode
    });

    postBook(json)
      .then(response => response.json())
      .then(responseJson => {
        console.log(responseJson);
        const book = {
          id: responseJson.id,
        };
        navigate('EntryTags', { book });

      })
      .catch(error => {
        console.warn(error);
        postBook(json)
          .then(response => response.json())
          .then(responseJson => {
            console.log(responseJson);
            const book = {
              id: responseJson.id,
            };
            navigate('EntryTags', { book });
          })
          .catch(error => console.error(error));
      });
  };

  renderNoPermissions = () => {
    return (
      <View style={styles.noPermissions}>
        <Text style={styles.noPermissionsText}>
          カメラを使用できません
        </Text>
      </View>
    );
  };

  renderHeader = () => {
    return (
      <View style={styles.header}>
        <Text style={styles.stringWarn}>
          978
          <Text style={styles.text}>
            から始まるバーコードを画面に合わせてください
          </Text>
        </Text>
      </View>
    );
  };

  renderCamera = () => {
    return (
      <View style={styles.cameraContainer}>
        <BarCodeScanner
          style={styles.camera}
          onBarCodeRead={this._handleBarCodeRead}
        >
          <View style={styles.cameraInline}>
            <Text/>
          </View>
        </BarCodeScanner>
      </View>
    );
  };

  renderFooter = () => {
    let statusText = <View/>;
    if (this.state.status == 'ok') {
      statusText = <Text style={styles.statusOk}>読み取りました</Text>;
    } else if (this.state.status == 'invalid') {
      statusText = <Text style={styles.statusNo}>数字をお確かめください</Text>;
    }

    return (
      <View style={styles.footer}>
        <Button
          title={statusText}
          loading={(this.state.status == 'reading')}
          loadingProps={{ size: "large", color: "rgba(111, 202, 186, 1)" }}
          titleStyle={{ fontWeight: "700" }}
          clear={true}
        />
      </View>
    );
  };

  renderScanScreen = () => {
    const header = this.renderHeader();
    const camera = this.renderCamera();
    const footer = this.renderFooter();

    return (
      <View style={styles.cameraScreen}>
        {header}
        <Image
          style={styles.imageSize}
          source={require('./assets/ISBN_sample.png')}
        />
        {camera}
        {footer}
      </View>
    );
  };

  render() {
    const screen = this.state.permissionsGranted
      ? this.renderScanScreen()
      : this.renderNoPermissions();

    return (
      <View style={styles.container}>
        {screen}
      </View>
    );
  }
}

const { width } = Dimensions.get('window');

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000',
  },
  noPermissions: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  noPermissionsText: {
    color: 'white'
  },
  cameraScreen: {
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  header: {
    alignItems: 'center',
    margin: 5,
  },
  stringWarn: {
    color: '#cd5c5c',
  },
  text: {
    color: '#f3f3f3',
    textAlign: 'center',
  },
  imageSize: {
    width: 221 / 2,
    height: 93 / 2
  },
  body: {
    alignItems: 'center',
    margin: 5,
  },
  cameraContainer: {
    width: width * 2 / 3,
    height: width * 2 / 3,
    borderColor: '#f3f3f3',
    borderWidth: 1,
  },
  camera: {
    flex: 1,
  },
  cameraInline: {
    flex: 0.5,
    borderColor: '#cd5c5c',
    borderBottomWidth: 1,
  },
  footer: {
    margin: 5,
  },
  statusOk: {
    textAlign: 'center',
    fontSize: 30,
    color: '#3eb370'
  },
  statusNo: {
    textAlign: 'center',
    fontSize: 30,
    color: '#e95464'
  },
});
