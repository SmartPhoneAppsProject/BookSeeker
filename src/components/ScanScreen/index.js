import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Image,
  ActivityIndicator,
} from 'react-native';
import {
  BarCodeScanner,
  Permissions,
} from 'expo';
import { Text } from 'react-native-elements';

import { index as styles } from './Styles';

export default class ScanScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isbn: null,
      permissionsGranted: false,
      status: 'reading',
    };
  }

  async componentWillMount() {
    const { status } = await Permissions.askAsync(Permissions.CAMERA);
    this.setState({ permissionsGranted: status === 'granted' });
  }

  handleBarCodeRead = ({ type, data }) => {
    const { navigation } = this.props;

    console.log(data);
    if (BarCodeScanner.Constants.BarCodeType.ean13 === type) {
      if (data.slice(0, 3) === '978') { // ISBNを読み取ったとき
        if (this.state.isbn !== data) {
          this.setState({
            isbn: data,
            status: 'ok',
          });
          setTimeout(() => {
            const isbn = String(parseInt(data, 10));
            navigation.navigate('EntryTags', {
              title: this.props.title,
              image: this.props.image.base64,
              published: this.props.published,
              isbn,
            });
          }, 1000);
        }
      } else { // バーコードであるがISBNでないとき
        this.setState({ status: 'invalid' });
      }
      setTimeout(() => {
        this.setState({ status: 'reading' });
      }, 1000);
    } else { // バーコードでないとき
      this.setState({ status: 'reading' });
    }
  };

  renderNoPermissions = () => (
    <View style={styles.noPermissions}>
      <Text style={styles.noPermissionsText}>
        カメラを使用できません
      </Text>
    </View>
  );

  renderHeader = () => (
    <View style={styles.header}>
      <Text style={styles.stringWarn}>
        978
        <Text style={styles.text}>
          から始まるバーコードを画面に合わせてください
        </Text>
      </Text>
    </View>
  );

  renderCamera = () => (
    <View style={styles.cameraContainer}>
      <BarCodeScanner
        style={styles.camera}
        onBarCodeRead={this.handleBarCodeRead}
      >
        <View style={styles.cameraInline}>
          <Text />
        </View>
      </BarCodeScanner>
    </View>
  );

  renderFooter = () => {
    let statusText = <ActivityIndicator size="large" />;
    if (this.state.status === 'ok') {
      statusText = <Text h4 style={styles.statusOk}>読み取りました</Text>;
    } else if (this.state.status === 'invalid') {
      statusText = <Text h4 style={styles.statusNo}>数字をお確かめください</Text>;
    }

    return (
      <View style={styles.footer}>
        {statusText}
      </View>
    );
  };

  renderScanScreen = () => {
    const header = this.renderHeader();
    const camera = this.renderCamera();
    const footer = this.renderFooter();

    return (
      <View style={styles.cameraScreen}>
        {header}
        <Image
          style={styles.imageSize}
          source={require('../../../assets/ISBN_sample.png')}
        />
        {camera}
        {footer}
      </View>
    );
  };

  render() {
    const screen = this.state.permissionsGranted
      ? this.renderScanScreen()
      : this.renderNoPermissions();

    return (
      <View style={styles.container}>
        {screen}
      </View>
    );
  }
}

ScanScreen.navigationOptions = {
  title: 'バーコードリーダー',
};

ScanScreen.propTypes = {
  title: PropTypes.string.isRequired,
  published: PropTypes.string.isRequired,
  image: PropTypes.string,
};

ScanScreen.defaultProps = {
  image: '',
};
