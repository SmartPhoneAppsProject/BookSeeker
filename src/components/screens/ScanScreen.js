import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  Image,
  Dimensions,
  ActivityIndicator,
} from 'react-native';
import {
  BarCodeScanner,
  Permissions,
} from 'expo';
import { Text } from 'react-native-elements';

import { postBook } from '../../utils/Network';

export default class ScanScreen extends Component {
  static navigationOptions = {
    title: 'バーコードリーダー',
  };

  constructor(props) {
    super(props);
    this.state = {
      isbn: null,
      permissionsGranted: false,
      status: 'reading',
    };
  }

  async componentWillMount() {
    const { status } = await Permissions.askAsync(Permissions.CAMERA);
    this.setState({ permissionsGranted: status === 'granted' });
  }

  handleBarCodeRead = ({ type, data }) => {
    console.log(data);
    if (BarCodeScanner.Constants.BarCodeType.ean13 === type) {
      if (data.slice(0, 3) === '978') { // ISBNを読み取ったとき
        if (this.state.isbn !== data) {
          this.setState({
            isbn: data,
            status: 'ok',
          });
          setTimeout(() => {
            const isbn = parseInt(data, 10);
            this.registerBook(isbn);
          }, 1000);
        }
      } else { // バーコードであるがISBNでないとき
        this.setState({ status: 'invalid' });
      }
      setTimeout(() => {
        this.setState({ status: 'reading' });
      }, 1000);
    } else { // バーコードでないとき
      this.setState({ status: 'reading' });
    }
  };

  registerBook = (isbn) => {
    const { navigation } = this.props;
    const { params } = this.props.navigation.state;

    const json = JSON.stringify({
      title: params.title,
      image: params.photo.base64,
      published: params.published,
      isbn,
    });

    postBook(json)
      .then(response => response.json())
      .then((responseJson) => {
        console.log(responseJson);
        const book = {
          id: responseJson.id,
        };
        navigation.navigate('EntryTags', { book });
      })
      .catch((error) => {
        console.warn(error);
        postBook(json)
          .then(response => response.json())
          .then((responseJson) => {
            console.log(responseJson);
            const book = {
              id: responseJson.id,
            };
            navigation.navigate('EntryTags', { book });
          })
          .catch(e => console.error(e));
      });
  };

  renderNoPermissions = () => (
    <View style={styles.noPermissions}>
      <Text style={styles.noPermissionsText}>
        カメラを使用できません
      </Text>
    </View>
  );

  renderHeader = () => (
    <View style={styles.header}>
      <Text style={styles.stringWarn}>
        978
        <Text style={styles.text}>
          から始まるバーコードを画面に合わせてください
        </Text>
      </Text>
    </View>
  );

  renderCamera = () => (
    <View style={styles.cameraContainer}>
      <BarCodeScanner
        style={styles.camera}
        onBarCodeRead={this.handleBarCodeRead}
      >
        <View style={styles.cameraInline}>
          <Text />
        </View>
      </BarCodeScanner>
    </View>
  );

  renderFooter = () => {
    let statusText = <ActivityIndicator size="large" />;
    if (this.state.status === 'ok') {
      statusText = <Text h4 style={styles.statusOk}>読み取りました</Text>;
    } else if (this.state.status === 'invalid') {
      statusText = <Text h4 style={styles.statusNo}>数字をお確かめください</Text>;
    }

    return (
      <View style={styles.footer}>
        {statusText}
      </View>
    );
  };

  renderScanScreen = () => {
    const header = this.renderHeader();
    const camera = this.renderCamera();
    const footer = this.renderFooter();

    return (
      <View style={styles.cameraScreen}>
        {header}
        <Image
          style={styles.imageSize}
          source={require('../../../assets/ISBN_sample.png')}
        />
        {camera}
        {footer}
      </View>
    );
  };

  render() {
    const screen = this.state.permissionsGranted
      ? this.renderScanScreen()
      : this.renderNoPermissions();

    return (
      <View style={styles.container}>
        {screen}
      </View>
    );
  }
}

const { width } = Dimensions.get('window');

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#000',
  },
  noPermissions: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  noPermissionsText: {
    color: 'white',
  },
  cameraScreen: {
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  header: {
    alignItems: 'center',
    margin: 5,
  },
  stringWarn: {
    color: '#cd5c5c',
  },
  text: {
    color: '#f3f3f3',
    textAlign: 'center',
  },
  imageSize: {
    width: 221 / 2,
    height: 93 / 2,
  },
  body: {
    alignItems: 'center',
    margin: 5,
  },
  cameraContainer: {
    width: width * (2 / 3),
    height: width * (2 / 3),
    borderColor: '#f3f3f3',
    borderWidth: 1,
  },
  camera: {
    flex: 1,
  },
  cameraInline: {
    flex: 0.5,
    borderColor: '#cd5c5c',
    borderBottomWidth: 1,
  },
  footer: {
    margin: 5,
  },
  statusOk: {
    textAlign: 'center',
    fontSize: 30,
    color: '#3eb370',
  },
  statusNo: {
    textAlign: 'center',
    fontSize: 30,
    color: '#e95464',
  },
});
