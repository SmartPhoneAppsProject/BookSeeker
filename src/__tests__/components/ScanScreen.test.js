import React from 'react';
import { shallow } from 'enzyme';
import ScanScreen from '../../components/ScanScreen';
import _books from '../../api/books.json';

describe('<ScanScreen />', () => {
  it('正しくレンダリングされていること', () => {
    const book = _books[0];
    const wrapper = shallow(<ScanScreen
      title={book.title}
      published={book.published}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
