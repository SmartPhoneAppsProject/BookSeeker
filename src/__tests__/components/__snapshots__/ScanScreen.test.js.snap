// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScanScreen /> 正しくレンダリングされていること 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ScanScreen
    image=""
    published="2018-02-09T00:00:00Z"
    title="javascript "
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <TextElement
          style={
            Object {
              "color": "white",
            }
          }
        >
          カメラを使用できません
        </TextElement>
      </View>,
      "style": Object {
        "backgroundColor": "#000",
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <TextElement
          style={
            Object {
              "color": "white",
            }
          }
        >
          カメラを使用できません
        </TextElement>,
        "style": Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "padding": 10,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "カメラを使用できません",
          "style": Object {
            "color": "white",
          },
        },
        "ref": null,
        "rendered": "カメラを使用できません",
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "padding": 10,
            }
          }
        >
          <TextElement
            style={
              Object {
                "color": "white",
              }
            }
          >
            カメラを使用できません
          </TextElement>
        </View>,
        "style": Object {
          "backgroundColor": "#000",
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <TextElement
            style={
              Object {
                "color": "white",
              }
            }
          >
            カメラを使用できません
          </TextElement>,
          "style": Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "padding": 10,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "カメラを使用できません",
            "style": Object {
              "color": "white",
            },
          },
          "ref": null,
          "rendered": "カメラを使用できません",
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
